packages/connect/src/createScClient/createScClient.ts

```diff
@@ -11,19 +20,293 @@ export interface ScClient {
+      // messages around until we receive the subscription-id message.
+      if (!this.#orphanMessages.has(subscriptionId))
+        this.#orphanMessages.set(subscriptionId, [])
+      this.#orphanMessages.get(subscriptionId)!.push(decodedResponse)
+    }
+
+    this.#chain = this.#getChain(onResponse).then((chain) => {
+      hc.setSendJsonRpc(chain.sendJsonRpc)
+
+      this.#isChainReady = false
+      const cleanup = () => {
+        // If there are any callbacks left, we have to reject/error them.
+        // Otherwise, that would cause a memory leak.
+        const disconnectionError = new Error("Disconnected")
+        this.#requests.forEach((cb) => cb(disconnectionError))
+        this.#subscriptions.forEach(([cb]) => cb(disconnectionError))
+        this.#subscriptions.clear()
+        this.#orphanMessages.clear()
+      }
+
+      let staleUnsubscriptionMessages: string[] = []
+      const killStaleSubscriptions = () => {
+        staleUnsubscriptionMessages.forEach((json) => {
+          try {
+            // it's a fire-and-forget request we don't care about the answer
+            hc.sendJsonRpc(json)
+          } catch (_) {}
+        })
+      }
+
+      hc.start((health) => {
+        const isReady =
+          !health.isSyncing && (health.peers > 0 || !health.shouldHavePeers)
+
+        // if it's the same as before, then nothing has changed and we are done
+        if (this.#isChainReady === isReady) return
+
+        if (!isReady) {
+          // If we've reached this point, that means that the chain used to be "ready"
+          // and now we are about to emit `disconnected`.
+          //
+          // This will cause the PolkadotJs API think that the connection is
+          // actually dead. In reality the smoldot chains is not dead, of course.
+          // However, we have to cleanup all the existing callbacks because when
+          // the smoldot chain stops syncing, then we will emit `connected` and
+          // the PolkadotJs API will try to re-create the previous
+          // subscriptions and requests.
+          //
+          // However, when the smoldot chain is no longer syncing, then we will
+          // have to let smoldot know that those stale subscriptions are no longer
+          // needed. that's why -before we perform the cleanup of `this.#subscriptions`-
+          // we prepare the json messages that we will have to send to the smoldot
+          // chain to kill those stale subscriptions.
+          staleUnsubscriptionMessages = [...this.#subscriptions.values()].map(
+            ([, { method, id }]) => this.#coder.encodeJson(method, [id]),
+          )
+          cleanup()
+        } else killStaleSubscriptions()
+
+        this.#isChainReady = isReady
+        this.#eventemitter.emit(isReady ? "connected" : "disconnected")
+      })
+
+      return {
+        ...chain,
+        sendJsonRpc: hc.sendJsonRpc.bind(hc),
+        remove: () => {
```

Comment:
It's really confusing to also name this `remove`. I have no idea which `remove` you're calling where.

Reply:
It's a way to ensure that whenever the consumer of the exposed `chain` calls `remove` and/or `sendJsonRpc` that the right logic takes place... So, we ware making the consumer unaware of the "magic" that goes behind enhancing the `Chain`.

In other words: this code is mostly making up for the convoluted API of the `healthChecker`. You see, the `healthChecker` is in fact a leaky "Chain enhancer". If the `healthChecker` had had a cleaner API, like this  one:

```ts
type HealthHandler = (healthUpdate: SmoldotHealth) => void              
type HealthChain = Chain & {                                            
  onHealthUpdate: HealthHandler                                         
  getCurrentHealth: () => SmoldotHealth                                 
}                                                                       
type GetHealthChain = (handler: JsonRpcCallback) => Promise<HealthChain>
                                                                        
type GetChain = (handler: JsonRpcCallback) => Promise<Chain>            
                                                                        
type HealthChecker = (getChain: GetChain) => GetHealthChain             
```

or like this one:

```ts
type HealthHandler = (healthUpdate: SmoldotHealth) => void  
type GetHealthChain = (                                     
  handler: JsonRpcCallback,                                 
  healthHandler: HealthHandler,                             
) => Promise<Chain>                                         
                                                            
type GetChain = (handler: JsonRpcCallback) => Promise<Chain>
                                                            
type HealthChecker = (getChain: GetChain) => GetHealthChain 
```

Then I wouldn't have to "hack" the exposed chain that much.

That being said: I'm open to suggestions. Is there a concrete action that you want me to take here?